### PR TITLE
bazel_7: 7.4.0 -> 7.4.1

### PR DESCRIPTION
**Pull request by boltzmannrain**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by boltzmannrain** [comment]
@ofborg build bazel_7

**Comment by boltzmannrain** [comment]
@ofborg build bazel_7

**Comment by FlyingWombat** [comment]
build fails at patchPhase.
```
bazel> Running phase: unpackPhase
bazel> unpacking source archive /nix/store/9dh7hw9x3jw35s0xjb2zinlkh3ckr0yw-bazel-7.4.1-dist.zip
...
bazel> applying patch /nix/store/9857sd4kzzpl1xcx81hwh98ldma58ya3-nix-hacks.patch
bazel>   --- src/main/java/com/google/devtools/build/lib/analysis/CommandHelper.java.bak-nix    1980-01-01 00:00:00.000000000 +0000
bazel>   +++ src/main/java/com/google/devtools/build/lib/analysis/CommandHelper.java    2024-11-13 23:44:49.445124439 +0000
bazel>   @@ -111 +111 @@
bazel>   -   * Maximum total command-line length, in bytes, not counting "/bin/bash -c ".
bazel>   +   * Maximum total command-line length, in bytes, not counting "/nix/store/fhb4h7i7dq3lfp5lgbr6xb7pvlzmcd5w-bash/bin/bash -c ".
bazel>   @@ -114 +114 @@
bazel>   -   * For short commands, we just use /bin/bash -c command.
bazel>   +   * For short commands, we just use /nix/store/fhb4h7i7dq3lfp5lgbr6xb7pvlzmcd5w-bash/bin/bash -c command.
bazel>   @@ -327 +327 @@
bazel>   -    // Use vanilla /bin/bash for actions running on mac machines.
bazel>   +    // Use vanilla /nix/store/fhb4h7i7dq3lfp5lgbr6xb7pvlzmcd5w-bash/bin/bash for actions running on mac machines.
bazel>   @@ -329 +329 @@
bazel>   -        ? PathFragment.create("/bin/bash")
bazel>   +        ? PathFragment.create("/nix/store/fhb4h7i7dq3lfp5lgbr6xb7pvlzmcd5w-bash/bin/bash")
bazel> patching file src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java
bazel> Hunk #1 FAILED at 178.
bazel> Hunk #2 succeeded at 634 (offset 24 lines).
bazel> 1 out of 2 hunks FAILED -- saving rejects to file src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java.rej
```

**Comment by boltzmannrain** [comment]
Looks like build with `enableNixHacks=true` was broken on 7.3.1 update too, and there's no test on `bazel_7` itself that checks that. The only package using `buildBazelPackage` with `bazel = bazel_7` seems to be `fcitx5-mozc`.

So that's indeed a regression but small-scale. Patch needs to be updated, originally the conflicting part about markers seems to be coming from https://github.com/NixOS/nixpkgs/pull/186106 Not yet sure what's it should do and how to do it on latest bazel codebase, will try to have a look

**Comment by boltzmannrain** [comment]
And looks like the marker hack is required for `buildBazelPackage` to work, removing that patch yields
```
error: builder for '/nix/store/rbjk3i01a5np4zi3s21w138agwrgn2ff-fcitx5-mozc-2.30.5544.102.drv' failed with exit code 1;
       last 10 log lines:
       >   /build/source/src/WORKSPACE.bazel:44:13: in <toplevel>
       > Repository rule http_archive defined at:
       >   /build/output/external/bazel_tools/tools/build_defs/repo/http.bzl:387:31: in <toplevel>
       > ERROR: /build/output/external/bazel_tools/tools/build_defs/repo/http.bzl:136:45: An error occurred during the fetch of repository 'bazel_skylib':
       >    Traceback (most recent call last):
       >     File "/build/output/external/bazel_tools/tools/build_defs/repo/http.bzl", line 136, column 45, in _http_archive_impl
       >           download_info = ctx.download_and_extract(
       > Error in download_and_extract: java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.4.1/bazel-skylib-1.4.1.tar.gz, https://github.com/bazelbuild/bazel-skylib/releases/download/1.4.1/bazel-skylib-1.4.1.tar.gz] to /build/output/external/bazel_skylib/temp15662579802079223249/bazel-skylib-1.4.1.tar.gz: Unknown host: github.com
       > ERROR: Error computing the main repository mapping: no such package '@@bazel_skylib//': java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.4.1/bazel-skylib-1.4.1.tar.gz, https://github.com/bazelbuild/bazel-skylib/releases/download/1.4.1/bazel-skylib-1.4.1.tar.gz] to /build/output/external/bazel_skylib/temp15662579802079223249/bazel-skylib-1.4.1.tar.gz: Unknown host: github.com
       > 
       For full logs, run 'nix-store -l /nix/store/rbjk3i01a5np4zi3s21w138agwrgn2ff-fcitx5-mozc-2.30.5544.102.drv'.
```

**Comment by boltzmannrain** [comment]
But if we also remove removal of .marker files from buildBazelPackage deps outputs it seems to be building fcitx5-mozc
https://github.com/boltzmannrain/nixpkgs/commit/14d715d10896051c42359923f5e0d930e112dbce on this commit build succeeded, maybe we don't need that hack anymore?

**Comment by RossComputerGuy** [comment]
Result of `nixpkgs-review pr 355723` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>fcitx5-mozc</li>
    <li>fcitx5-mozc-ut</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bazel_7</li>
  </ul>
</details>

**Comment by boltzmannrain** [comment]
This PR itself doesn't fix fcitx5-mozc build (buildBazelBackage with bazel=bazel-7), I don't yet fully understand the patch used there to update it.

Found a little more background on the markers hack here https://github.com/NixOS/nixpkgs/pull/65374

It seems that removing a section of failing patch + some more changes may work for fcitx5-mozc as in https://github.com/boltzmannrain/nixpkgs/commit/14d715d10896051c42359923f5e0d930e112dbce#diff-fd4e3dd34e0c6b6d8da7ac1dc25530696d40844760fd3aebf103e2444c55c6a8R163

But I don't know if that change to buildBazelPackage setup works for other packages, maybe for bazel 5/6 we should keep existing behavior of removing markers to be on the safe side (building tensorflow might be a good test)

Additionally if we decide to change buildBazelPackage behavior it probably makes sense to invalidate fixed output derivations, similarly how they get invalidated on bazel version changes in 
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/build-bazel-package/default.nix#L219

**Comment by musjj** [comment]
> This PR itself doesn't fix fcitx5-mozc build

Yeah, it looks like the problem is the patch from `buildBazelPackage`. It's just patch failure, so you just need to adjust and create a separate patch for `7.x.x`, but I guess you also need to add a version check to make sure the right patch gets applied to the right bazel version.

**Review by musjj — CHANGES_REQUESTED** [review]
Can you try this patch:

<details>
<summary><code>nix-hacks-7.patch</code></summary>

```diff
diff --git a/src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java b/src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java
index 53e6494..09052cd 100644
--- a/src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java
+++ b/src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java
@@ -55,6 +55,7 @@ import com.google.devtools.build.skyframe.SkyFunctionException.Transience;
 import com.google.devtools.build.skyframe.SkyKey;
 import com.google.devtools.build.skyframe.SkyValue;
 import java.io.IOException;
+import java.util.Collections;
 import java.util.Map;
 import java.util.Optional;
 import java.util.TreeMap;
@@ -193,18 +194,8 @@ public final class RepositoryDelegatorFunction implements SkyFunction {
       }
 
       if (shouldUseCachedRepos(env, handler, repoRoot, rule)) {
-        // Make sure marker file is up-to-date; correctly describes the current repository state
-        byte[] markerHash = digestWriter.areRepositoryAndMarkerFileConsistent(handler, env);
-        if (env.valuesMissing()) {
-          return null;
-        }
-        if (markerHash != null) { // repo exist & up-to-date
-          return RepositoryDirectoryValue.builder()
-              .setPath(repoRoot)
-              .setDigest(markerHash)
-              .setExcludeFromVendoring(excludeRepoFromVendoring)
-              .build();
-        }
+        // Nix hack: Always consider cached dirs as up-to-date
+        return RepositoryDirectoryValue.builder().setPath(repoRoot).setDigest(digestWriter.writeMarkerFile(Collections.emptyMap())).build();
       }
 
       /* At this point: This is a force fetch, a local repository, OR The repository cache is old or
@@ -634,11 +625,12 @@ public final class RepositoryDelegatorFunction implements SkyFunction {
         builder.append(escape(key)).append(" ").append(escape(value)).append("\n");
       }
       String content = builder.toString();
-      try {
-        FileSystemUtils.writeContent(markerPath, UTF_8, content);
-      } catch (IOException e) {
-        throw new RepositoryFunctionException(e, Transience.TRANSIENT);
-      }
+      // Nix hack: Do not write these pesky marker files
+      //try {
+      //  FileSystemUtils.writeContent(markerPath, UTF_8, content);
+      //} catch (IOException e) {
+      //  throw new RepositoryFunctionException(e, Transience.TRANSIENT);
+      //}
       return new Fingerprint().addString(content).digestAndReset();
     }
 
diff --git a/src/main/java/com/google/devtools/build/lib/shell/JavaSubprocessFactory.java b/src/main/java/com/google/devtools/build/lib/shell/JavaSubprocessFactory.java
index 649647c..64d05b5 100644
--- a/src/main/java/com/google/devtools/build/lib/shell/JavaSubprocessFactory.java
+++ b/src/main/java/com/google/devtools/build/lib/shell/JavaSubprocessFactory.java
@@ -165,7 +165,6 @@ public class JavaSubprocessFactory implements SubprocessFactory {
     }
     builder.command(argv);
     if (params.getEnv() != null) {
-      builder.environment().clear();
       builder.environment().putAll(params.getEnv());
     }
```
</details>

You can conditionally import it just for `7.x.x` like this:

```nix
# See enableNixHacks argument above.
++ lib.optional enableNixHacks (
  if versionAtLeast version "7" then ./nix-hacks-7.patch else ./nix-hacks.patch
);
```

I'm not really sure if the `Collections.emptyMap()` hack here have any serious implications, but fcitx5 does build now.

The problem was that we need to get the recorded inputs via `readMarkerFile` and `markerPath` now, but they're both private. Maybe we could patch `areRepositoryAndMarkerFileConsistent` to make sure it always succeeds, but I'm not sure if it'd have any unwanted side effects.

**Comment by RossComputerGuy** [comment]
Please don't use substring to check versions, we have `lib.versionAtLeast` and `lib.versionOlder`.

**Comment by FlyingWombat** [comment]
Do we still need the patch? fcit5-mozc is the only package explicitly depending on bazel_7, and the build succeeds with this commit that removes the patch and .marker file removal. https://github.com/boltzmannrain/nixpkgs/commit/14d715d10896051c42359923f5e0d930e112dbce

We could try migrating other packages to bazel_7 to see if it works, but they might break on bazel_7 regardless.

**Comment by musjj** [comment]
> Do we still need the patch?

I'm not really sure either, but I've had experiences where bazel builds ok once, but fails in subsequent builds (due to hash changes). The PR you linked seems to imply that.

Can you try clearing your cache and rebuilding it to see if it works?

> fcit5-mozc is the only package explicitly depending on bazel_7

Yes for now. fcitx5-mozc doesn't actually need bazel 7 right now, but it will be a hard requirement for both fcitx5-mozc and mozc on the next version bump (because it uses `local.bzl`).

File: pkgs/development/tools/build-managers/bazel/bazel_7/nix-build-bazel-package-hacks.patch [hunk 1]
```diff
@@ -1,46 +1,37 @@
...
-+        return RepositoryDirectoryValue.builder().setPath(repoRoot).setDigest(digestWriter.writeMarkerFile()).build();
-       }
- 
-       /* At this point: This is a force fetch, a local repository, OR The repository cache is old or
-@@ -610,11 +600,12 @@ public final class RepositoryDelegatorFunction implements SkyFunction {
-         builder.append(escape(key)).append(" ").append(escape(value)).append("\n");
+-              .setExcludeFromVendoring(excludeRepoFromVendoring)
++              .setDigest(digestWriter.writeMarkerFile(Collections.emptyMap()))
+               .build();
+         }
```
Comment (musjj) [inline pkgs/development/tools/build-managers/bazel/bazel_7/nix-build-bazel-package-hacks.patch:31]: Have you tested if the patch works? It seems to generate an extra scope:

```java
if (shouldUseCachedRepos(env, handler, repoRoot, rule)) {
  {
    // Nix hack: Always consider cached dirs as up-to-date
    return RepositoryDirectoryValue.builder()
        .setPath(repoRoot)
        .setDigest(digestWriter.writeMarkerFile(Collections.emptyMap()))
        .build();
  }
}
```

Does this actually return from the function or just from the scope (not really familiar with Java).

Comment (boltzmannrain) [reply pkgs/development/tools/build-managers/bazel/bazel_7/nix-build-bazel-package-hacks.patch:31]: Extra nested block creates a nested lexical scope but `return` still refers to the enclosing function as there's no block-level returns. Added it to reduce formatting indentation diff against bazel upstream

**Review by musjj — APPROVED** [review]
Ok LGTM. It should be fine to downgrade bazel for `fcitx5-mozc` for now.

**Comment by philiptaron** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 355723`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>fcitx5-mozc</li>
    <li>fcitx5-mozc-ut</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>bazel_7</li>
  </ul>
</details>

**Review by philiptaron — APPROVED** [review]

**Comment by philiptaron** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 355723`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>bazel_7</li>
  </ul>
</details>

**Pull request merged**